### PR TITLE
use a script instead of piping to stdin in run_docker_build

### PR DESCRIPTION
**Pull request by minrk**
Puts the script in build_artefacts/conda-forge-build.sh for use in the image.

It would appear that some tests interact with the stdin pipe, causing run_docker_build to finish early with apparent success, skipping the upload step. They seem to be doing this by consuming or detaching stdin somehow.

By using a regular script, pipes aren't involved, fixing at least one known cause of tests preventing upload.

Further, future similar failures should be protected against by writing a canary file in the build_artefacts dir at the end of the script. This is checked after `docker run`, in case other bugs reintroduce early exit with apparent success.

This is confirmed to fix the failure I was seeing on the new petsc-feedstock. The problematic test involved `mpiexec`, if that's relevant.

Other possibly affected cases:
- https://github.com/conda-forge/erlang-feedstock/issues/1
- https://github.com/conda-forge/libarchive-feedstock/issues/3

### CODE REVIEW

**Comment by jakirkham** [comment]
Good stuff @minrk ! Thanks for putting this together. 😀

Have definitely been suspicious of the pipe redirection for a while. I believe part of @pelson 's motivation for that design was to be security conscious. So we will need to evaluate these changes from that standpoint. Though maybe there are other concerns that he can share.

We should probably do something similar at staged-recipes. Though before we go down that road let me take a closer look at this implementation. Expect I should have some time later next week.

**Comment by jakirkham** [comment]
Also do you have any links to builds, commits and/or PRs that are relevant from the petsc feedstock?

**Comment by minrk** [comment]
This build was run with the re-render and re-enabling the problematic test: https://circleci.com/gh/minrk/petsc-feedstock/3

This build is otherwise identical, rendered with smithy-1.3.3: https://circleci.com/gh/minrk/petsc-feedstock/4

Both *pass*, but note how the latter doesn't show the "No BINSTAR_TOKEN present..." output of run_docker_build.sh, indicating that it stopped at conda-build.

**Comment by minrk** [comment]
I'm not sure that writing a script in the feedstock can introduce a new security issue, because run_test.sh, etc. are mounted from the same directory and can be modified (after launching the build, even) to change what code the container runs. Although I guess now there is a file containing the BINSTAR_TOKEN, which may be what was meant to be avoided. I'm not sure how to avoid that with the script.

Things I wasn't so sure about:

- where to put the build script / canary. I used build_artefacts because it's already git-ignored, but maybe it should be a new ignored folder. I aimed for minimal new things to start in a relatively unfamiliar codebase.
- whether the canary is really necessary - I would be surprised if the issue is ever reintroduced, so maybe the canary check is

File: conda_smithy/templates/run_docker_build.tmpl [hunk 1]
```diff
@@ -52,5 +53,18 @@ upload_or_check_non_existence /recipe_root {{ owner }} --channel={{ channel }} |
 {%- endfor -%}
 {%- endblock -%}
 
+# signal build-done with a file to protect against
+# an early exit with status=0 appearing to have succeeded
+touch /feedstock_root/build_artefacts/conda-forge-build-done
 EOF
 
+docker run -i \
```
Comment (jakirkham) [inline conda_smithy/templates/run_docker_build.tmpl]: Given we are giving up redirection, what do you think about enabling TTY? Have already had a few issues where not having a TTY has caused problems (e.g. `ncurses`).

**Comment by jakirkham** [comment]
> Although I guess now there is a file containing the BINSTAR_TOKEN, which may be what was meant to be avoided. I'm not sure how to avoid that with the script.

Yes, this is what I meant. Sorry if that was cryptic. Not sure ATM either, but will try to think on it.

**Comment by jakirkham** [comment]
> where to put the build script / canary. I used build_artefacts because it's already git-ignored, but maybe it should be a new ignored folder. I aimed for minimal new things to start in a relatively unfamiliar codebase.

My inclination would be to put it in `ci_support` directory. We can also just add the file to `.gitignore`.

File: conda_smithy/templates/run_docker_build.tmpl (hunk 1 above)
Comment (minrk) [inline conda_smithy/templates/run_docker_build.tmpl]: sure, makes sense

**Comment by minrk** [comment]
We can pass environment variables to docker on the CLI with `-e` and through files with `--env-file`, but both seem similarly snoopable. I don't see a way to tell `docker run` to 'use my environment variable' without echoing it somewhere other than the pipe.

Two pipey solutions if the file is to be avoided:

1. write the script via pipe to `docker build -` so we get a docker image with the token-having script
2. keep the pipe as-is, but write the whole content to a script in the container before running it

Both of these would avoid the token at rest on the host filesystem, but snoops with access to the docker service would be able to reach in and retrieve it

**Comment by jakirkham** [comment]
This needs a rebase. Probably broken by PR ( https://github.com/conda-forge/conda-smithy/pull/285 ). Should be an easy fix.

File: conda_smithy/templates/run_docker_build.tmpl (hunk 1 above)
Comment (minrk) [inline conda_smithy/templates/run_docker_build.tmpl]: TTY enabled

**Comment by minrk** [comment]
Sorry for forgetting about this one. Rebased and TTY enabled, per comment above.

**Comment by jakirkham** [comment]
Thanks @minrk . I do want to take a closer look at this again, but it may not be for a couple weeks due to the other big changes with CI providers and `conda-build`.

**Comment by jakirkham** [comment]
Tentatively scheduling this for 2.1.0. I know we still have some more discussion and review to do, but am hopeful we should get this resolved soon. Please let me know if you think this is reasonable.

**Comment by jakirkham** [comment]
@gqmelo , I'd be curious to hear your thoughts on this change. Would you have a chance to take a look?

**Comment by gqmelo** [comment]
I like the idea, but I'm sorry for my ignorance: is there a good reason to keep only one template? I mean, couldn't we have a ```build.tmpl``` so that conda smithy would render a ```ci_support/build.sh``` and ```ci_support/run_docker_build.sh```? What would be the implications of rendering a ```ci_support/build.sh``` and commit it?

It is just that this would be simpler and make it easy to use ```build.sh``` without ```docker```. I've needed this sometimes to investigate some errors that happened inside docker but not on my native machine (but then I had to extract only the conda-build stuff from ```run_docker_build.sh``` to try to reproduce it).

Anyway this would be probably another issue. I think these changes are nice

**Comment by minrk** [comment]
> What would be the implications of rendering a ci_support/build.sh and commit it?

Exactly as it is, build.sh includes the BINSTAR_TOKEN env from circle, which we don't have at render time. I think it could be generated as a template and use an env-file to pass the token, though.

**Comment by minrk** [comment]
Actually, I might have a better idea - if we can do the upload from the host environment instead of the build environment, this should be even simpler, and not have any issues with tokens.

**Comment by gqmelo** [comment]
> Actually, I might have a better idea - if we can do the upload from the host environment instead of the build environment, this should be even simpler, and not have any issues with tokens.

That would be nice but you would have to setup the host environment with dependencies needed by the upload script, right?

**Comment by dalcinl** [comment]
@minrk @jakirkham What about running docker two times? The first one you build using a script, the second one you upload passing the token through stdin. I think this is even better, as you never expose the token to the build scripts.

**Comment by jakirkham** [comment]
Was thinking about the same thing, @dalcinl.

**Comment by minrk** [comment]
@dalcinl good idea. This now runs docker twice:

1. run the build script with a plain script and proper tty (should fix mpiexec issues)
2. run the upload step in a second container, passing the token via stdin

**Comment by minrk** [comment]
This makes it easier to move build.sh out into a regular templated file, but I'd prefer to save that for a different PR and/or person, since it's adding new items in the template machinery that I have not dealt with before.

**Comment by minrk** [comment]
Another option that might be simpler is to use a docker `env-file` to pass the token, in which case it could go back to a single docker build script. This would still mean the token is on disk, but it could be brief (e.g. only for the duration of `docker create`). Upload already does this with the token, writing it to a file for `anaconda upload` to read it.

**Comment by dalcinl** [comment]
There is no need to write the token to a file. IIUC, `anacona -t <token> upload ...` can be passed either a token or a path to a file containing a token.

**Comment by dalcinl** [comment]
@minrk This is getting overly complicated. I'm not a docker expert, but I think you can implement this with just two ``docker run ...`` commands. As long as you mount the ``/feedstock_root`` volume in both invocations, the build artifacts from the build run will be available to the upload run. Am I missing something?

**Comment by minrk** [comment]
@dalcinl indeed it could be done in two docker commands (it can also be done in one if an env-file is acceptable). I did it this way to avoid duplicating the setup:

1. run the build
2. commit the build to an image
3. run the upload in that image

To do it in two `docker run` commands, the setup has to be duplicated.

I've switched it back to a single build with an env-file, which is the simplest, I think. The only change is that I used `create` and `start`, so that the env file could be deleted promptly, rather than waiting for build to finish.

**Comment by dalcinl** [comment]
@minrk I don't get why this would be more secure that ``docker run`` with an env file. Also, I do not like the fact that the token is available in the environment in the build stage. Imagine a leftover invokation to ``env`` (in an attempt to debug a build issue) in a recipe gets merged. Then we will be exposing the token to the world in our build logs.

**Comment by minrk** [comment]
@dalcinl it's not, it just reduces the time during which the file is on disk. I'm not sure this is something to be especially concerned about, because reading these files doesn't seem any harder than interrogating the environment variables of running processes via `/proc/pid/environ`.

If there is a desire to protect the binstar token from build, I think that's a separate issue because all the build scripts on all CI expose this.

**Comment by jakirkham** [comment]
Yep, we had a discussion a while ago about improving handling of uploads. One idea that @pelson had was to have an intermediate service we upload packages to. That service could validate a number of things before actually doing the upload. In the interim, we try to do our best here.

ref: https://conda-forge.hackpad.com/conda-forge-upload-service-N5evEX7bZAf

**Comment by jakirkham** [comment]
TBH the biggest concern in my mind is someone accidentally exposing the token. For example, someone is debugging a build and decides to run `env` to figure out what is going wrong. If we can make it so it is extremely unlikely the token shows up there, we should be in a good place.

**Comment by minrk** [comment]
@jakirkham would you like me to try to do that here, or should it be a separate task since any change here would only affect circle builds?

**Comment by jakirkham** [comment]
Sorry @minrk that this has been neglected. We should revive it, but probably not until after we switch to `conda-build` 3, which should hopefully be soon.

**Comment by minrk** [comment]
Rebased on master

**Comment by minrk** [comment]
rebased. This has gotten simpler since the original PR since the matrix env stuff is no longer needed.

**Comment by ocefpaf** [comment]
Ouch. Sorry @minrk another PR that we dropped the ball.

File: conda_smithy/templates/run_docker_build.tmpl [hunk 2]
```diff
@@ -60,17 +58,33 @@ conda install --yes --quiet conda-forge-ci-setup=1
...
-test -f "$FEEDSTOCK_ROOT/build_artifacts/conda-forge-build-done" || exit 1
 
+# Create and run the container in two commands,
+# so the token env isn't on disk too long
+
+CONTAINER=$(docker create -t \
+           -v "${RECIPE_ROOT}":/home/conda/recipe_root \
+           -v "${FEEDSTOCK_ROOT}":/home/conda/feedstock_root \
+           --env-file ${FEEDSTOCK_ROOT}/env \
+           -e HOST_USER_ID="${HOST_USER_ID}" \
```
Comment (isuruf) [inline conda_smithy/templates/run_docker_build.tmpl]: You've missed the CONFIG env variable here.

File: conda_smithy/templates/run_docker_build.tmpl [hunk 3]
```diff
@@ -60,17 +58,33 @@ conda install --yes --quiet conda-forge-ci-setup=1
...
 {%- endfor %}
 
 touch /home/conda/feedstock_root/build_artifacts/conda-forge-build-done
 EOF
 
-# double-check that the build got to the end
-# see https://github.com/conda-forge/conda-smithy/pull/337
-# for a possible fix
+set +x
+echo "BINSTAR_TOKEN=${BINSTAR_TOKEN}" > "${FEEDSTOCK_ROOT}/env"
```
Comment (isuruf) [inline conda_smithy/templates/run_docker_build.tmpl]: Put this inside .circleci folder?

File: conda_smithy/templates/run_docker_build.tmpl (hunk 2 above)
Comment (minrk) [reply conda_smithy/templates/run_docker_build.tmpl]: #692

File: conda_smithy/templates/run_docker_build.tmpl [hunk 4]
```diff
@@ -60,17 +58,33 @@ conda install --yes --quiet conda-forge-ci-setup=1
...
 set -x
-test -f "$FEEDSTOCK_ROOT/build_artifacts/conda-forge-build-done" || exit 1
 
+# Create and run the container in two commands,
+# so the token env isn't on disk too long
+
+CONTAINER=$(docker create -t \
+           -v "${RECIPE_ROOT}":/home/conda/recipe_root \
+           -v "${FEEDSTOCK_ROOT}":/home/conda/feedstock_root \
+           --env-file ${FEEDSTOCK_ROOT}/env \
```
Comment (jakirkham) [inline conda_smithy/templates/run_docker_build.tmpl]: Was not aware of this option. Nice fix!

File: conda_smithy/templates/run_docker_build.tmpl (hunk 3 above)
Comment (jakirkham) [reply conda_smithy/templates/run_docker_build.tmpl]: Actually would it make sense to save this in a directory that is not mounted in the container? To propose an arbitrary option, what about `~/.conda-smithy/upload.token`?

**Comment by isuruf** [comment]
This doesn't work in circleci and hangs with no output after the build.sh is run.

**Comment by jakirkham** [comment]
We can change back to `docker run` and just point to an env file outside the feedstock.

**Comment by isuruf** [comment]
I'm going to revert it as I can't find anyway to fix it

**Comment by jakirkham** [comment]
That seems reasonable, @isuruf. FWIW I do want us to solve this issue. Just think trying to tackle it in `conda-smithy` 3 when we are very close to releasing (and thus getting conda-build 3 support), is a bit ambitious personally.

**Comment by minrk** [comment]
FWIW, after the single typo fixed in #692 I can confirm that this PR results in working builds with no issues. [successful build here](https://circleci.com/gh/minrk/testpath-feedstock/tree/smithy-test).

**Pull request merged**